cmd/end2endtest/ballot.go

```diff
@@ -63,48 +103,186 @@ func (*E2EBallotElection) Teardown() error {
 }
 
 func (t *E2EBallotElection) Run() error {
-	bdata := ballotData{
-		maxValue:     t.election.TallyMode.MaxValue,
-		maxCount:     t.election.TallyMode.MaxCount,
-		maxTotalCost: t.election.TallyMode.MaxTotalCost,
-		costExponent: t.election.TallyMode.CostExponent,
-	}
+	nvotes := t.elections[0].config.nvotes
+	var wg sync.WaitGroup
+	errCh := make(chan error, 4)
 
-	choices, expectedResults := ballotVotes(bdata, len(t.voterAccounts))
+	choicesE1, expectedResults := ballotVotes(t.elections[0].election.TallyMode, nvotes, "ranked", true)
+	choicesE2, expectedResults2 := ballotVotes(t.elections[1].election.TallyMode, nvotes, "quadratic", false)
+	choicesE3, expectedResults3 := ballotVotes(t.elections[2].election.TallyMode, nvotes, "range", true)
+	choicesE4, expectedResults4 := ballotVotes(t.elections[3].election.TallyMode, nvotes, "approval", false)
```

Comment:
please avoid hardcoded strings, ballotVotes is an unexported func so you control both caller and func, and can replace those strings by some global `const` that are much easier to debug when refactoring

Reply:
this is a general suggestion for you to keep in mind when coding.

in this particular case i'll suggest to really refactor the whole PR, that would also make these 4 strings disappear